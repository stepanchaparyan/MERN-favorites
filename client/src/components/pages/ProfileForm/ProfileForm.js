import React, { useState, useContext, useEffect } from 'react';
import { useIntl } from 'react-intl';
import ProfileContext from '../../../context/profileContext/profileContext';
import {
  Container,
  Module,
  Form,
  Input,
  Select,
  Option,
  DefaultOption,
  ProfileData,
  Text,
  DatePickerStyled
} from './ProfileFormStyled';
import localization from './localization';
import { FORM } from '../../../constants';

const { INPUT, SELECT, TEL_PLACEHOLDER, TEL_PATTERN } = FORM;

const ProfileForm = () => {
  const context = useContext(ProfileContext);
  const { formatMessage } = useIntl();

  const { updateProfile, editProfile, toggle_Form, toggleForm } = context;

  const [newProfile, setProfile] = useState(editProfile);

  useEffect(() => {
    setProfile(editProfile);
  }, [context]);

  const { name, surname, email, gender, birthDay, phoneNumber } = newProfile;

  const [birthDate, setBirthDate] = useState(Date.parse(birthDay));

  const setNewDate = date => {
    setBirthDate(date);
    setProfile({
      ...newProfile,
      birthDay: date
    });
  };

  const onchange = e => {
    setProfile({
      ...newProfile,
      [e.target.name]: e.target.value
    });
  };

  const onsubmit = e => {
    e.preventDefault();
    updateProfile(newProfile);
    toggle_Form(!toggleForm);
  };

  const cancelEdit = () => {
    toggle_Form(!toggleForm);
  };

  return (
    <Container>
      <Module>
        <Form onSubmit={onsubmit}>
          <ProfileData>
            <Text>{formatMessage(localization.name)}:</Text>
            <Input
              type={INPUT.TYPE.TEXT}
              placeholder={formatMessage(localization.name)}
              name={INPUT.NAME.NAME}
              value={name}
              onChange={onchange}
            />
          </ProfileData>
          <ProfileData>
            <Text>{formatMessage(localization.surname)}:</Text>
            <Input
              type={INPUT.TYPE.TEXT}
              placeholder={formatMessage(localization.surname)}
              name={INPUT.NAME.SURNAME}
              value={surname}
              onChange={onchange}
            />
          </ProfileData>
          <ProfileData>
            <Text>{formatMessage(localization.email)}:</Text>
            <Input
              type={INPUT.TYPE.EMAIL}
              placeholder={formatMessage(localization.email)}
              name={INPUT.NAME.EMAIL}
              value={email}
              onChange={onchange}
            />
          </ProfileData>
          <ProfileData>
            <Text>{formatMessage(localization.gender)}:</Text>
            <Select value={gender} name={SELECT.NAME.GENDER} onChange={onchange}>
              <DefaultOption value={SELECT.VALUES.OTHER}>
                {formatMessage(localization.selectCategory)}
              </DefaultOption>
              <Option value={SELECT.VALUES.MALE}>{formatMessage(localization.male)}</Option>
              <Option value={SELECT.VALUES.FEMALE}>{formatMessage(localization.female)}</Option>
              <Option value={SELECT.VALUES.OTHER}>{formatMessage(localization.other)}</Option>
            </Select>
          </ProfileData>
          <ProfileData>
            <Text>{formatMessage(localization.birthday)}:</Text>
            <DatePickerStyled
              selected={birthDate}
              onChange={date => setNewDate(date)}
              peekNextMonth
              showMonthDropdown
              showYearDropdown
              dropdownMode="select"
            />
          </ProfileData>
          <ProfileData>
            <Text>{formatMessage(localization.phone)}:</Text>
            <Input
              type={INPUT.TYPE.TEL}
              placeholder={TEL_PLACEHOLDER}
              pattern={TEL_PATTERN}
              name={INPUT.NAME.PHONE_NUMBER}
              value={phoneNumber}
              onChange={onchange}
            />
          </ProfileData>
          <Input type={INPUT.TYPE.SUBMIT} value={formatMessage(localization.update)} />
          <Input
            onClick={cancelEdit}
            type={INPUT.TYPE.BUTTON}
            value={formatMessage(localization.cancel)}
          />
        </Form>
      </Module>
    </Container>
  );
};
export default ProfileForm;
