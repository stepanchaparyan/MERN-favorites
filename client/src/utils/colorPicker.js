export const colorPicker = category => {
  let color;
  switch (category) {
    case 'Film':
      color = 'red';
      break;
    case 'Music':
      color = 'blue';
      break;
    case 'Books':
      color = 'limeGreen';
      break;
    case 'Other':
      color = 'orange';
      break;
    default:
      color = 'black';
  }
  return color;
};
