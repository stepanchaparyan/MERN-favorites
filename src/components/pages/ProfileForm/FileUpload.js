import React, { useState, useContext, useEffect } from 'react';
import Progress from './Progress';
import ProfileContext from '../../../context/profileContext/profileContext';
import {
  Container,
  Input,
  InputHidden,
  ProfileImage,
  LabelEdit,
  LabelUpload,
  IconEdit,
  Filename,
  customStyles,
  ButtonConfirm,
  ButtonClose,
  ButtonCancel,
  ModalContainer,
  ModalTitleContainer,
  ModalTitle,
  ModalTextContainer,
  ModalButtonsContainer
} from './FileUploadStyled';
import EditIcon from '../../../assets/icon-edit.png';
import UploadIcon from '../../../assets/icon-upload.png';
import Modal from 'react-modal';

const FileUpload = () => {
  const context = useContext(ProfileContext);
  const {
    updateProfile,
    editProfile,
    update_File,
    uploadedFile,
    uploadPercentage,
    message
  } = context;
  const [newProfile, setProfile] = useState(editProfile);

  useEffect(() => {
    setProfile(editProfile);
  }, []);

  const { image } = newProfile;
  const [file, setFile] = useState('');
  const [filename, setFilename] = useState('');
  const [modalIsOpen, setIsOpen] = useState(false);

  const onChange = e => {
    setFile(e.target.files[0]);
    setFilename(e.target.files[0].name);
  };

  const onSubmit = e => {
    e.preventDefault();
    const formData = new FormData();
    formData.append('file', file);
    update_File(formData);
  };

  const setImagePath = () => {
    setProfile({
      ...newProfile,
      image: filename
    });
    openModal(true);
  };

  const onChangePath = () => {
    updateProfile(newProfile);
  };

  Modal.setAppElement('#root');

  const openModal = () => {
    setIsOpen(true);
  };

  const closeModal = () => {
    setIsOpen(false);
  };

  return (
    <Container>
      <form onSubmit={onSubmit}>
        <ProfileImage src={require(`../../../assets/${image}`)} />
        <LabelEdit htmlFor="file-edit">
          <IconEdit src={EditIcon}></IconEdit>
        </LabelEdit>
        <InputHidden id="file-edit" type="file" onChange={onChange} />
        <Filename>{filename}</Filename>

        {filename && (
          <>
            <Progress percentage={uploadPercentage} />
            <LabelUpload htmlFor="file-upload">
              <IconEdit src={UploadIcon}></IconEdit>Upload
            </LabelUpload>
            <InputHidden id="file-upload" type="submit" />
          </>
        )}
      </form>
      {uploadedFile && <Input type="button" value="Update Image" onClick={setImagePath} />}
      <Modal
        isOpen={modalIsOpen}
        onRequestClose={closeModal}
        style={customStyles}
        contentLabel="Modal"
      >
        <ModalContainer>
          <ModalTitleContainer>
            <ModalTitle>Confirm</ModalTitle>
            <ButtonClose onClick={closeModal}>X</ButtonClose>
          </ModalTitleContainer>
          <ModalTextContainer></ModalTextContainer>
          <ModalButtonsContainer>
            <ButtonConfirm onClick={onChangePath}>Confirm</ButtonConfirm>
            <ButtonCancel onClick={closeModal}>Cancel</ButtonCancel>
          </ModalButtonsContainer>
        </ModalContainer>
      </Modal>
    </Container>
  );
};

export default FileUpload;
