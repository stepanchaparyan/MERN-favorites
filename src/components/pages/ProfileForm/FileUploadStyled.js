import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 10px;
  min-width: 200px;
`;

export const Input = styled.input`
  border: 0px;
  width: 200px;
  padding: 8px 10px;
  margin: 4px;
  outline: none;
  border-radius: 4px;

  &[type='submit'] {
    width: 50%;
    border: 0;
    background-color: ${props => props.theme.color};
    font-weight: 600;
    font-size: 16px;
    margin-top: 20px;
    border-radius: 5%;
    cursor: pointer;
    margin-bottom: 10px;
  }
  &[type='button'] {
    width: 70%;
    border: 0;
    background-color: ${props => props.theme.color};
    font-weight: 600;
    font-size: 16px;
    border-radius: 5%;
    cursor: pointer;
    margin: 10px;
  }
`;

export const InputHidden = styled.input`
  display: none;
`;

export const ProfileImage = styled.img`
  width: 80px;
  height: 80px;
  border-radius: 50%;
`;

export const IconEdit = styled.img`
  width: 20px;
  height: 20px;
  padding: 16px 5px 0 0;
`;

export const Button = styled.button`
  width: 35%;
  border: 0;
  background-color: gray;
  font-weight: 600;
  font-size: 16px;
  border-radius: 5%;
  cursor: pointer;
  margin-bottom: 10px;
`;

export const LabelEdit = styled.label`
  cursor: pointer;
`;

export const LabelUpload = styled.label`
  cursor: pointer;
  color: black;
  font-size: 21px;
  border: 1px black;
  background-color: lightgrey;
  border-radius: 3px;
  padding: 3px 6px 0 6px;
`;

export const Filename = styled.div`
  font-size: 12px;
  margin: 12px 0;
`;

export const customStyles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    display: 'flex',
    flexDirection: 'column',
    padding: '0',
    width: '350px',
    height: '200px'
  }
};

export const ButtonConfirm = styled.button`
  border: 0;
  background-color: ${props => props.theme.color};
  font-weight: 600;
  font-size: 16px;
  border-radius: 5%;
  cursor: pointer;
  margin: 12px;
  padding: 8px;
`;

export const ButtonClose = styled.button`
  background-color: ${props => props.theme.color};
  border: 0;
  font-weight: 600;
  font-size: 16px;
  border-radius: 5%;
  cursor: pointer;
  margin-right: 8px;
`;

export const ButtonCancel = styled.button`
  width: 30%;
  border: 0;
  background-color: lightgrey;
  font-weight: 600;
  font-size: 16px;
  border-radius: 5%;
  cursor: pointer;
  margin: 12px;
  padding: 8px;
`;

export const ModalContainer = styled.div`
  background-color: #f8f8f8;
  display: flex;
  flex-direction: column;
`;

export const ModalTitleContainer = styled.div`
  background-color: ${props => props.theme.color};
  display: flex;
  justify-content: space-between;
`;

export const ModalTitle = styled.div`
  font-size: 20px;
  padding: 16px;
  font-weight: bold;
`;

export const ModalTextContainer = styled.div`
  display: flex;
  min-height: 80px;
`;

export const ModalButtonsContainer = styled.div`
  display: flex;
  justify-content: center;
  border-top: 1px solid lightGray;
  width: 94%;
  margin: 0 3%;
`;
