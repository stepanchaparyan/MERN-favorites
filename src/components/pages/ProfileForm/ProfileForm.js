import React, { useState, useContext, useEffect } from 'react';
import ProfileContext from '../../../context/profileContext/profileContext';
import {
  Container,
  Module,
  ProfileImage,
  Form,
  Input,
  Select,
  Option,
  DefaultOption,
  ProfileData,
  Text,
  DatePickerStyled
} from './ProfileFormStyled';
import Img from '../../../assets/elephant.png';

const ProfileForm = () => {
  const context = useContext(ProfileContext);

  const { updateProfile, editProfile, toggle_Form, toggleForm } = context;

  const [newProfile, setProfile] = useState(editProfile);

  useEffect(() => {
    setProfile(editProfile);
  }, [context]);

  const { name, surname, email, gender, birthDay, phone } = newProfile;

  const [birthDate, setBirthDate] = useState(Date.parse(birthDay));

  const setNewDate = date => {
    setBirthDate(date);
    setProfile({
      ...newProfile,
      birthDay: date
    });
  };

  const onchange = e => {
    setProfile({
      ...newProfile,
      [e.target.name]: e.target.value
    });
  };

  const onsubmit = e => {
    e.preventDefault();
    updateProfile(newProfile);
    toggle_Form(!toggleForm);
  };

  const cancelEdit = () => {
    toggle_Form(!toggleForm);
  };

  return (
    <Container>
      <Module>
        <ProfileImage src={Img}></ProfileImage>
        <Form onSubmit={onsubmit}>
          <ProfileData>
            <Text>Name:</Text>
            <Input type="text" placeholder="Name" name="name" value={name} onChange={onchange} />
          </ProfileData>
          <ProfileData>
            <Text>Surname:</Text>
            <Input
              type="text"
              placeholder="Surname"
              name="surname"
              value={surname}
              onChange={onchange}
            />
          </ProfileData>
          <ProfileData>
            <Text>Email:</Text>
            <Input
              type="email"
              placeholder="Email"
              name="email"
              value={email}
              onChange={onchange}
            />
          </ProfileData>
          <ProfileData>
            <Text>Gender:</Text>
            <Select value={gender} name="gender" onChange={onchange}>
              <DefaultOption value="Other">Select category</DefaultOption>
              <Option value="male">Male</Option>
              <Option value="female">Female</Option>
              <Option value="other">Other</Option>
            </Select>
          </ProfileData>
          <ProfileData>
            <Text>BirthDay:</Text>
            <DatePickerStyled
              selected={birthDate}
              onChange={date => setNewDate(date)}
              peekNextMonth
              showMonthDropdown
              showYearDropdown
              dropdownMode="select"
            />
          </ProfileData>
          <ProfileData>
            <Text>Phone:</Text>
            <Input
              type="tel"
              placeholder="93123456"
              pattern="[0-9]{8}"
              name="phone"
              value={phone}
              onChange={onchange}
            />
          </ProfileData>
          <Input type="submit" value={'Update'} />
          <Input onClick={cancelEdit} type="button" value="Cancel" />
        </Form>
      </Module>
    </Container>
  );
};
export default ProfileForm;
