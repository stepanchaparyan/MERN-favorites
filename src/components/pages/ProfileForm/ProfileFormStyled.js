import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const Container = styled.div`
  display: flex;
  justify-content: center;
  color: ${props => props.theme.color};
  padding: 0 32px 50px 32px;
  font-weight: 600;
  font-size: 24px;
  background-color: lightgray;
  border-radius: 4px;
`;

export const Module = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 16px;
  width: 50%;
`;

export const WelcomeText = styled.div`
  font-size: 40px;
  padding: 16px;
`;

export const LinkStyled = styled(Link)`
  color: ${props => props.theme.color};
  margin: 32px;
  text-decoration: none;
  font-size: 32px;
`;

export const Logo = styled.img`
  height: 200px;
  position: absolute;
  bottom: 80px;
  right: 60px;
`;

export const LoadingMessage = styled.h3`
  display: flex;
  justify-content: center;
  color: ${props => props.theme.color};
`;

export const TextData = styled.div`
  background: whiteSmoke;
  border: 1px solid cadetBlue;
  width: 85%;
`;

export const ProfileData = styled.div`
  display: flex;
  font-size: 20px;
  padding: 8px;
  width: 100%;
`;

export const Text = styled.div`
  color: black;
  width: 45%;
`;

export const Data = styled.div`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 200px;
  width: 50%;
`;

export const ProfileImage = styled.img`
  height: 50px;
`;

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Input = styled.input`
  border: 0px;
  width: 200px;
  padding: 8px 10px;
  margin: 4px;
  outline: none;
  border-radius: 4px;

  &[type='submit'] {
    width: 50%;
    border: 0;
    background-color: ${props => props.theme.color};
    font-weight: 600;
    font-size: 16px;
    margin-top: 20px;
    border-radius: 5%;
    cursor: pointer;
    margin-bottom: 10px;
  }

  &[type='button'] {
    width: 35%;
    border: 0;
    background-color: gray;
    font-weight: 600;
    font-size: 16px;
    border-radius: 5%;
    cursor: pointer;
    margin-bottom: 10px;
  }
`;

export const Select = styled.select`
  width: 220px;
  background: white;
  height: 32px;
  outline: none;
  border: none;
  border-radius: 4px;
  padding-left: 5px;
  margin: 4px;
`;

export const Option = styled.option`
  color: black;
  height: 47px;
  padding: 20px;
  font-size: 18px;
  color: ${props => props.theme.color};
`;

export const DefaultOption = styled.option`
  display: none;
`;
