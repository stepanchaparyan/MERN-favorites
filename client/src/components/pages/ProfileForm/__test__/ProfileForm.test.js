import React from 'react';
import { renderSnapshot } from '../../../../utils/tests';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import ProfileForm from '../ProfileForm.js';
import { Form, DatePickerStyled, Input, CancelButton } from '../ProfileFormStyled';

jest.mock('react-intl', () => ({
  useIntl: jest.fn(() => ({ formatMessage: m => m.defaultMessage })),
  defineMessages: i => i
}));

jest.mock('react', () => {
  const ActualReact = jest.requireActual('react');
  return {
    ...ActualReact,
    useEffect: jest.fn().mockImplementation(f => f),
    useContext: jest.fn(() => ({
      editProfile: {
        name: 'testName',
        surname: 'testSurname',
        email: 'test@email.com',
        gender: 'testGender',
        birthDay: 'testBirtday',
        phoneNumber: '123456789'
      },
      toggle_Form: jest.fn(),
      updateProfile: jest.fn()
    }))
  };
});

describe('ProfileForm component test with Enzyme', () => {
  test('renders snapshot', () => {
    renderSnapshot(ProfileForm);
  });

  test('should click on Input elements', () => {
    const component = shallow(<ProfileForm />);
    const inputs = component.find(Input);
    inputs.forEach(input => {
      input.simulate('change', { target: { value: 'target' } });
      expect();
    });
  });

  test('should click on Form element', () => {
    const component = shallow(<ProfileForm />);
    const form = component.find(Form);
    const preventDefault = jest.fn();
    form.simulate('submit', { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
  });

  test('should click on DatePicker element', () => {
    const component = shallow(<ProfileForm />);
    const datePicker = component.find(DatePickerStyled);
    datePicker.simulate('change');
  });

  test('should click on CancelButton element', () => {
    const component = shallow(<ProfileForm />);
    const cancelButton = component.find(CancelButton);
    cancelButton.simulate('click');
  });
});
