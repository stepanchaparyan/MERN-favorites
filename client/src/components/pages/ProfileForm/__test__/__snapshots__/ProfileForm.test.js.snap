// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm component test with Enzyme renders snapshot 1`] = `
<styled.div>
  <styled.div>
    <styled.form
      onSubmit={[Function]}
    >
      <styled.div>
        <styled.div>
          Name
          :
        </styled.div>
        <styled.input
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value="testName"
        />
      </styled.div>
      <styled.div>
        <styled.div>
          Surname
          :
        </styled.div>
        <styled.input
          name="surname"
          onChange={[Function]}
          placeholder="Surname"
          type="text"
          value="testSurname"
        />
      </styled.div>
      <styled.div>
        <styled.div>
          Email
          :
        </styled.div>
        <styled.input
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value="test@email.com"
        />
      </styled.div>
      <styled.div>
        <styled.div>
          Gender
          :
        </styled.div>
        <styled.select
          name="gender"
          onChange={[Function]}
          value="testGender"
        >
          <styled.option
            value="Other"
          >
            Select Category
          </styled.option>
          <styled.option
            value="male"
          >
            Male
          </styled.option>
          <styled.option
            value="female"
          >
            Female
          </styled.option>
          <styled.option
            value="Other"
          >
            Other
          </styled.option>
        </styled.select>
      </styled.div>
      <styled.div>
        <styled.div>
          BirthDay
          :
        </styled.div>
        <Styled(o)
          dropdownMode="select"
          onChange={[Function]}
          peekNextMonth={true}
          selected={NaN}
          showMonthDropdown={true}
          showYearDropdown={true}
        />
      </styled.div>
      <styled.div>
        <styled.div>
          Phone
          :
        </styled.div>
        <styled.input
          name="phoneNumber"
          onChange={[Function]}
          pattern="[0-9]{8}"
          placeholder="93123456"
          type="tel"
          value="123456789"
        />
      </styled.div>
      <styled.button>
         
        Update
      </styled.button>
      <styled.button
        onClick={[Function]}
      >
        Cancel
      </styled.button>
    </styled.form>
  </styled.div>
</styled.div>
`;
